File: 백준/Silver/5568. 카드 놓기/카드 놓기.cc
#include <bits/stdc++.h>
#include<unordered_map>
#define MAX 987654321
#define MIN -987654321
#define MOD 1000000007
#define pb push_back
#define pob pop_back
using namespace std;
typedef long long int ll;
typedef unsigned long long int ull;
typedef pair<int, int> pii;
typedef pair<int, string> pis;
typedef vector<vector<int>> vvec;
typedef vector<int> vec;
typedef vector<pii> vec2;
typedef vector<vector<pii>> gra;
int N, M, T;
ll res;
int L;
int num;
int V, E, K;
vec v;
vec v2;
bool c[11];
set<string>s;
void solve(int idx,int cnt)
{
	if (cnt == M)
	{
		string str = "";
		for (auto cur : v2)
		{
			str = str + to_string(v[cur]);
		}
		s.insert(str);
		return;
	}
	for (int i = 0; i < N; i++)
	{
		if (!c[i])
		{
			c[i] = 1;
			v2.push_back(i);
			solve(i + 1, cnt + 1);
			v2.pop_back();
			c[i] = 0;
		}

	}
}
int main()
{
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
	cin >> N >> M;
	int i;
	for (i = 0; i < N; i++)
	{
		cin >> num;
		v.push_back(num);
	}
	solve(0,0);
	cout << s.size();
}

File: 백준/Silver/5568. 카드 놓기/카드 놓기.py
import sys
#sys.setrecursionlimit(100000)
def input(): return sys.stdin.readline().rstrip()
MAX = sys.maxsize
MOD = int(1e9)
def is_inside(x, y, N, M):
    return 0 <= x < N and 0 <= y < M
# dx=[0,1,0,-1,-1,-1,1,1]
# dy=[1,0,-1,0,-1,1,1,-1]
dx = [0, 0, 1,-1]
dy = [1, -1, 0, 0]
N=int(input())
K=int(input())
a=[int(input()) for _ in range(N)]
st=set()
c=[0 for _ in range(N+1)]
def go(idx,cnt,num):
    if cnt==K:
        st.add(num)
        return
    if idx==N:
        return
    for i in range(N):
        if c[i]:continue
        c[i]=1
        go(idx+1,cnt+1,num+str(a[i]))
        c[i]=0
go(0,0,"")
print(len(st))